_posts/2018-07-10-react-native-wdio.md

```diff
@@ -0,0 +1,358 @@
+the app installation taking a long time.
+
+There are various timeouts in play here at each level of tech stack: WebDriverIO,
+Mocha/Jasmine, Appium, and SauceLabs/TestObject. In our case, simple WebDriverIO tests
+kept failing intermittently until we completely understood the timeouts and tweaked them.
+
+**WebDriverIO timeouts:**
+- `connectionRetryTimeout`: Http request timeouts while trying to connect to Appium server.
+- `waitforTimeout`: Timeout for all `waitFor` commands in the tests.
+
+**TestFramework timeout:**
+A test is declared as a failure if it isn't completed within a certain time. Test frameworks
+and their respective timeout options are listed below.
+```js
+// Mocha
+mochaOpts: {
+  timeout
+}
+
+// Jasmine
+jasmineNodeOpts: {
+  defaultTimeoutInterval
+}
+
+// Cucumber
+cucumberOpts: {
+  timeout
+}
+```
+
+**Appium timeouts:**
+The following timeouts deal with how long Appium should wait for Android Virtual Device (AVD)
+emulator or iOS simulator to be ready and for the app to be installed.
+- `appWaitDuration`
+- `deviceReadyTimeout`
+- `androidDeviceReadyTimeout` (Android only)
+- `androidInstallTimeout` (Android only)
+- `avdLaunchTimeout` (Android only)
+- `avdReadyTimeout` (Android only)
+- `launchTimeout` (iOS only)
+
+- `newCommandTimeout` - Limits how long (in seconds) Appium will wait for a new command
+from the client before assuming the client quit
+
+**SauceLabs/TestObject timeouts:**
+- `commandTimeout`: Similar to Appium's newCommandTimeout, but for SauceLabs. How long
+Selenium can take to run a command.
+- `idleTimeout`: Limits how long a browser can wait for a test to send a new command. 
+- `maxDuration`: Limit the total time taken to run a test.
+
+## Conclusion:
+I hope the above discussion helps you set up some infrastructure for running WebDriverIO
+UI tests locally and remotely for emulators and real devices. Occasionally, we did face
+few intermittent UI test failures due to the app being unresponsive after installation or
+network latency. We have addressed this issue by adding retry logic on top of the
+WebDriverIO command to re-run the failed test suites. 90% of the time these tests pass
+on the second run. We are continuing to write UI tests using WebDriverIO + Appium for native
+which turned out to be pretty effective for developement teams while working with web and
+native.
+
+GoDaddy is looking for a full-stack mobile engineer to join our next generation Customer
+Experience mobile team in Kirkland. The team is building our next generation experiences
+for our small business customers to help them start, grow, and run their venture. If you
+have the passion, enthusiasm, and ability to create compelling interactions for customers
+on their way to making their small businesses great, we would like to talk to you! Apply
+here: https://careers.godaddy.com/job/cambridge/senior-software-engineer-mobile-focus/18045/8081067
```

Comment:
nit: make this a link so users don't have to copy + paste into the address bar

Reply:
Ah, check out how weird this comment looks because of those manual line breaks! It _renders_ fine into the blog because the HTML concatenates the lines, but it might be nice to people looking at the raw Markdown if the line breaks were removed so things can auto-wrap.